api-docs/activity/list_user_projects.yml

```diff
@@ -0,0 +1,178 @@
+endpoint: /api/users/<username>/projects
+method: GET
+title: List all projects that a user has signed up for
+description: Returns a JSON list of all projects that a user has signed up for, with details about the completion status
+authenticated: true
+request:
+  path_parameters:
+    - name: username
+      type: string
+      description: The unique username to use to fetch details about a user. Username is case-insensitive.
+      example: eva
+      required: true
+  query_parameters: []
+  body: {}
+response:
+  - type: array[activity]
+    required: true
+    description: A list of activity objects (project and status information) for all the projects user has signed up for
+    example: |
+      [
+        {
+          "slug": "build-your-own-shell",
+          "title": "Build your own Shell",
+          "short_description": "A short description of the project, in Markdown format. This is displayed on the Project's card on the home page and dashboard.",
+          "description": "In this project, we will build a Unix shell from scratch.\n\nWe'll use the Python's `subprocess` library to build shell.\n\n# Learning outcomes\n- Unix\n- Python\n",
+          "is_active": true,
+          "tags": ["Python", "Unix"]
+          "status": "Completed",
+          "tasks": [
+            {
+              "slug": "stdin-and-stdout",
+              "position": 1,
+              "title": "Stdin and stdout",
+              "description": "Take input from stdin and simply echo it to stdout in a loop.",
+              "status": "completed",
+              "status_description": "Test case 1: echo without spaces - passed\nTest case 2: echo with whitespace - passed\n",
+            },
```

Comment:
I am a bit doubtful about this format.
The activity should maybe have the username also, and it might be okay to not include other details about the project besides the slug ID and completion status of project and each task.
This is also because the other `/api/activity` endpoint will have to return the username in some form, and as part of the same non-nested object would be better.

Reply:
You think of two different ways to represent a resource. A short representation or a teaser and a long representation or the detail. When a resource is being presented as part of a collection, it can just use the teaser instead of the detail.  Also, here you are really returns the project activity, not projects themselves.

```
        {
          "slug": "build-your-own-shell",
          "title": "Build your own Shell",
          "short_description": "A short description of the project, in Markdown format. This is displayed on the Project's card on the home page and dashboard.",
          "tags": ["Python", "Unix"]
          "status": "Completed",
          "taskCount": 8,
          "completed": 7,
          "progress_percentage": 87.5 
    }
```